### PR TITLE
NEON 48 - keyword text aggregation

### PR DESCRIPTION
**Pull request by jacklechnercaci**
Fixes: [NEON-48](https://nextcentury.atlassian.net/browse/NEON-48)

Three tests were added just like in the number aggregation. Two of which are in the ticket. The other one is specifying the clusters in the cluster clause.

Is there a reason this is called keyword or text aggregation? I don't see a use for the keyword part yet. If I missed some implementation, I'll go back and do that.

### CODE REVIEW

**Comment by ThomasSchellenbergNextCentury** [comment]
@jacklechnercaci In ES, "text" is used for sentences/paragraphs while "keyword" is used for strings that are single conceptual phrases but may contain spaces or other delimiters.  ES will automatically index "text" by splitting the string into individual words, but will index "keyword" by keeping the whole string intact.  For example, if you used "text" on a "company_name" field, it would split "Next Century" into two separate terms, "Next" and "Century", and searching for the term "Next Century" would return zero hits!

We followed this pattern in NUCLEUS since it made sense to us and other datastores have similar patterns (think VARCHAR vs TEXT in SQL).

**Comment by ThomasSchellenbergNextCentury** [comment]
OK, I know this PR builds off the previous PR, but I can't get the diff to just show the new changes, so I'll just wait until that PR is merged first.

File: server/src/main/java/com/ncc/neon/services/ClusterService.java [hunk 1]
```diff
@@ -31,25 +44,11 @@
...
-        } catch (Exception e) {
-            e.printStackTrace();
-        }
+        String fieldType = this.clusterClause.getFieldType();
 
-        this.clusterClause = null; // reset cluster clause
-        return tabularQueryResult;
-    }
+        // to keep the check simple, specialty cases are for text
+        boolean isText = (fieldType.equals("text") || fieldType.equals("keyword")) ? true : false;
```
Comment (ThomasSchellenbergNextCentury) [inline server/src/main/java/com/ncc/neon/services/ClusterService.java:50]: Thinking forward: You may want to change isText from a boolean to an Enum.  That may make it easier to handle dates and lat/lon points in the future tickets.

File: server/src/main/java/com/ncc/neon/services/ClusterService.java [hunk 2]
```diff
@@ -211,6 +260,66 @@ private void aggregateNumbersInNewData(String fieldNameKey, String aggregationNa
...
 
+    /**
+     * Converts the incoming text bin e.g. "a", "bd" into a number.
+     *
+     * @param text the bin to be converted
+     * @return a BigDecimal corresponding to the index of the bin
+     */
+    private BigDecimal convertTextBinToNumber(String text) {
+        final String ALPHABET = "abcdefghijklmnopqrstuvwxyz";
+        final int ALPHABET_LENGTH = 26;
```
Comment (ThomasSchellenbergNextCentury) [inline server/src/main/java/com/ncc/neon/services/ClusterService.java:271]: Would ALPHABET and ALPHABET_LENGTH be better scoped as static class variables?

**Comment by ThomasSchellenbergNextCentury** [comment]
@jacklechnercaci I'm getting a unit test error after a `clean build` (don't know why I haven't seen it before):

```
> Task :server:test

com.ncc.neon.controllers.QueryControllerTests > executeQueryTest FAILED
    java.lang.AssertionError at QueryControllerTests.java:120
        Caused by: java.lang.AssertionError at QueryControllerTests.java:120
```

The stack trace says there's a null pointer exception at `ClusterService.cluster(ClusterService.java:78)`.  I think it's because the DummyQueryAdapter's execute method returns silly data that invalidates the cluster method's otherwise perfectly acceptable assumptions.  Feel free to modify the DummyQueryAdapter's output if that's easier for you.

https://github.com/NextCenturyCorporation/nucleus-data-server/blob/master/common/src/main/java/com/ncc/neon/adapters/dummy/DummyQueryAdapter.java#L27-L38

File: server/src/main/java/com/ncc/neon/services/ClusterService.java [hunk 3]
```diff
@@ -31,25 +44,11 @@
      * @return the new clustered results
      */
     public TabularQueryResult cluster(TabularQueryResult tabularQueryResult) {
-        try {
-            if (this.clusterClause.getFieldType().equals("number")) {
-                return this.aggregateNumber(tabularQueryResult);
-            }
-        } catch (Exception e) {
-            e.printStackTrace();
-        }
```
Comment (ThomasSchellenbergNextCentury) [inline server/src/main/java/com/ncc/neon/services/ClusterService.java:40]: Looks like you accidentally got rid of the try/catch.

File: server/src/main/java/com/ncc/neon/services/ClusterService.java [hunk 4]
```diff
@@ -211,6 +260,66 @@ private void aggregateNumbersInNewData(String fieldNameKey, String aggregationNa
...
+     * @return a BigDecimal corresponding to the index of the bin
+     */
+    private BigDecimal convertTextBinToNumber(String text) {
+        final String ALPHABET = "abcdefghijklmnopqrstuvwxyz";
+        final int ALPHABET_LENGTH = 26;
+        int total = 0;
+        for (int i = text.length() - 1; i > -1; i--) {
+            char c = text.charAt(i);
+            total += (ALPHABET.indexOf(c) + 1) * Math.pow(ALPHABET_LENGTH, i);
+        }
```
Comment (ThomasSchellenbergNextCentury) [inline server/src/main/java/com/ncc/neon/services/ClusterService.java:276]: This is a heck of a fascinating way to handle this task!

File: server/src/main/java/com/ncc/neon/services/ClusterService.java (hunk 2 above)
Comment (jacklechnercaci) [reply server/src/main/java/com/ncc/neon/services/ClusterService.java]: Fixed.

File: server/src/main/java/com/ncc/neon/services/ClusterService.java (hunk 1 above)
Comment (jacklechnercaci) [reply server/src/main/java/com/ncc/neon/services/ClusterService.java:54]: I understand making fieldType into the FieldType enum, but I'm not following what to make isText into. Maybe I should just use the FieldType enum and do another compare?

Comment (ThomasSchellenbergNextCentury) [reply server/src/main/java/com/ncc/neon/services/ClusterService.java:54]: Currently this method handles two use cases: `isText=true` and `isText=false`.  If this method will ever need to handle more than two use cases in the future (like dates and lat/lon points), then you may want to use an Enum.  Just a suggestion.  You can worry about it when it becomes relevant, if that's easier.

Comment (jacklechnercaci) [reply server/src/main/java/com/ncc/neon/services/ClusterService.java:54]: I'm changing this on the date one.

**Pull request merged**